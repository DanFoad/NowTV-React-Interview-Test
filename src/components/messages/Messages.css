/**** MESSAGE CONTAINER ****/
.messages-container {
  display: flex;
  flex-flow: column nowrap;
  align-items: stretch;
  margin: auto;
  max-width: 860px;
  height: 100%;
  border: 1px solid #E0E0E0;
  border-top: 0;
  border-bottom: 0;
}

/**** MESSAGES ****/
.messages {
  flex-grow: 1;
  display: flex;
  flex-flow: column nowrap;
  align-items: stretch;
  height: 100%;
  margin: auto;
  overflow-y: auto;
  padding: 0 16px;
  margin-right: 4px;
}
.message {
  display: flex;
  flex-flow: column nowrap;
  align-items: flex-start;
  max-width: 80%;
  margin: 16px 0;
  flex-shrink: 0;
}
.message.current-user {
  align-items: flex-end;
  align-self: flex-end;
}
.message-content {
  display: flex;
  flex-flow: row nowrap;
  align-items: flex-start;
  color: rgba(0, 0, 0, 0.6);
}
.current-user .message-content {
  flex-direction: row-reverse;
}
.message-details {
  position: relative;
  flex-grow: 1;
  padding: 12px;
  background-color: #EFEFEF;
  border-radius: 4px;
  border-top-left-radius: 0;
}
.message-details:after {
  content: '';
  position: absolute;
  width: 12px;
  height: 16px;
  background-color: inherit;
  left: -12px;
  top: 0;
  display: block;
  border-top-left-radius: 2px;
}
.current-user .message-details {
  border-top-left-radius: 4px;
  border-top-right-radius: 0;
  background-color: #ddedfa;
}
.current-user .message-details:after {
  left: auto;
  right: -12px;
  border-top-right-radius: 2px;
}
.user-details__avatar {
  width: 32px;
  height: 32px;
  border-radius: 50%;
  border: 6px solid #FFFFFF;
  position: relative;
  z-index: 100;
  background-color: #FFFFFF;
}
.message-info {
  display: flex;
  flex-flow: row nowrap;
  align-items: center;
  padding-left: 40px;
}
.current-user .message-info {
  padding-left: 0;
  padding-right: 40px;
}
.user-details__name {
  color: #4A4A4A;
  font-size: 0.9em;
}
.user-details__name:after {
  content: '\00B7';
  display: inline-block;
  margin: 0 8px;
}
.message-details__timestamp {
  font-size: 0.8em;
  color: #888888;
  display: inline-block;
  position: relative;
  top: 1px;
}

/**** MESSAGES HEADER ****/
.messages-header {
  display: flex;
  align-items: center;
  flex-flow: row nowrap;
  height: 64px;
  border-bottom: 1px solid #E0E0E0;
}
.messages-header h2 {
  flex: 1 0 0;
  text-align: center;
  color: #4A4A4A;
  font-weight: 500;
  font-size: 1.1em;
}
.messages-header__avatars {
  flex: 1 0 0;
  display: flex;
  flex-flow: row nowrap;
  align-items: center;
  padding-left: 16px;
}
.messages-header__avatar {
  height: 22px;
  width: 22px;
  border: 2px solid #FFFFFF;
  border-radius: 50%;
  position: relative;
}
.messages-header__avatar:nth-child(2) { left: -6px; }
.messages-header__avatar:nth-child(3) { left: -12px; }
.messages-header__avatar:nth-child(4) { left: -18px; }
.messages-header__avatar:nth-child(5) { left: -24px; }
.messages-header__additional {
  position: relative;
  left: -20px;
  font-size: 0.9em;
  color: #888888;
}
.messages-header__settings {
  flex: 1 0 0;
  text-align: right;
  padding-right: 24px;
}
.messages-header__settings img {
  height: 24px;
  cursor: pointer;
}

/**** MESSAGES INPUT ****/
.messages-input {
  padding: 16px 16px 24px;
  position: relative;
}
.messages-input__textarea {
  width: 100%;
  height: 96px;
  border: 1px solid #E0E0E0;
  border-radius: 4px;
  background-color: #FFFFFF;
  color: #888888;
  padding: 8px;
  font-size: 0.9em;
}
.messages-input__textarea:focus {
  outline: none;
}
.messages-input__send {
  border: 0;
  background-color: #79afdb;
  color: #FFFFFF;
  position: absolute;
  right: 32px;
  bottom: 40px;
  display: inline-block;
  padding: 8px 16px;
  text-transform: uppercase;
  border-radius: 2px;
  cursor: pointer;
}
.messages-input__actions {
  position: absolute;
  left: 32px;
  bottom: 40px;
  display: flex;
  flex-flow: row nowrap;
  align-items: center;
}
.messages-input__actions img {
  display: inline-block;
  height: 20px;
  margin-right: 8px;
  cursor: pointer;
}

.messages::-webkit-scrollbar {
  width: 8px;
  background-color: rgba(0,0,0,0);
  border-radius: 100px;
  -webkit-border-radius: 100px;
}
.messages::-webkit-scrollbar:hover {
  background-color: rgba(0, 0, 0, 0.09);
}

.messages::-webkit-scrollbar-thumb:vertical {
  background: rgba(0,0,0,0.5);
  border-radius: 100px;
  -webkit-border-radius: 100px;
}
.messages::-webkit-scrollbar-thumb:vertical:active {
  background: rgba(0,0,0,0.61);
  border-radius: 100px;
  -webkit-border-radius: 100px;
}

/**** LOADING ****/
.loading {
  display: flex;
  width: 100%;
  height: 100%;
  flex-flow: row nowrap;
  align-items: center;
  justify-content: center;
  margin: auto;
}