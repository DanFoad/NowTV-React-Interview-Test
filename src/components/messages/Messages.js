import React from 'react'
import PropTypes from 'prop-types'

import './Messages.css'

class Messages extends React.Component {
    componentDidMount() {
        this.refs.container.scrollTop = this.refs.container.scrollHeight
    }

    componentDidUpdate(prevProps, prevState) {
        this.refs.container.scrollTop = this.refs.container.scrollHeight
    }

    getUser(userId) {
        for (var i = 0; i < this.props.users.length; i++) {
            if (this.props.users[i].id === userId) {
            return this.props.users[i]
            }
        }
        return false
    }

    formatTimestamp(timestamp) {
        var datetime = new Date(Date.parse(timestamp))
        return datetime.toLocaleString()
    }

    getLoading() {
        return <div className="loading"><h1>Loading...</h1></div>
    }

    getSortedMessages() {
        let messages = [ ...this.props.messages ]
        messages.sort((a, b) => {
            var aTime = Date.parse(a.timestamp)
            var bTime = Date.parse(b.timestamp)
            return aTime - bTime
        })
        return messages
    }

    getMessages() {
        if (this.props.messages.length === 0)
            return this.getLoading()

        let messages = this.getSortedMessages()
        let messagesOutput = []

        for (var i in messages) {
            if (!messages.hasOwnProperty(i)) continue
            var message = messages[i]
            var user = this.getUser(message.userId)
            var messageClass = this.props.currentUser === message.userId ? "message current-user" : "message"
            if (!user) continue
            if (user.avatar == null) user.avatar = "/unknownuser.png"
            messagesOutput.push(
            <div key={message.id} className={ messageClass }>
                <div className="message-content">
                    <div className="user-details">
                        <img className="user-details__avatar" src={ user.avatar } alt="" />
                    </div>
                    <div className="message-details">
                        <p className="message-details__content" title={ user.email }>{ message.message }</p>
                    </div>
                </div>
                <div className="message-info">
                    <p className="user-details__name">{ user.firstName + " " + user.lastName }</p>
                    <p className="message-details__timestamp">{ this.formatTimestamp(message.timestamp) }</p>
                </div>
            </div>
            )
        }
        return messagesOutput
    }

    render() {
        return (
            <div className="messages" ref="container">
                { this.getMessages() }
            </div>
        )
    }
}

Messages.propTypes = {
    users: PropTypes.array,
    messages: PropTypes.array,
}

export default Messages