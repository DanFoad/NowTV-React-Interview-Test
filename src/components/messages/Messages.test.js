import React from "react";
import renderer from "react-test-renderer/";

import Messages from "./Messages";

function createNodeMock(element) {
  return {};
}

it("renders correctly", () => {
  const tree = renderer
    .create(
      <Messages
        messages={[
          {
            id: "cd445e6d-e514-424f-ba8f-16ec842002c6",
            userId: "fe27b760-a915-475c-80bb-7cdf14cc6ef3",
            message:
              "Quisque erat eros, viverra eget, congue eget, semper rutrum, nulla.",
            timestamp: "2017-02-09T04:27:38Z"
          }
        ]}
        users={[
          {
            id: "fe27b760-a915-475c-80bb-7cdf14cc6ef3",
            firstName: "Albert",
            lastName: "Rose",
            email: "arosec@bbb.org",
            avatar: "http://dummyimage.com/100x100.jpg/5fa2dd/ffffff",
            ip: "20.79.231.223"
          }
        ]}
        currentUser=""
      />,
      { createNodeMock }
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("doesn't display message if no user matches", () => {
  const tree = renderer
    .create(
      <Messages
        messages={[
          {
            id: "cd445e6d-e514-424f-ba8f-16ec842002c6",
            userId: "fe27b760-a915-475c-80bb-7cdf14cc6ef3",
            message:
              "Quisque erat eros, viverra eget, congue eget, semper rutrum, nulla.",
            timestamp: "2017-02-09T04:27:38Z"
          }
        ]}
        users={[]}
        currentUser=""
      />,
      { createNodeMock }
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders unknownuser avatar for users without avatar", () => {
  const tree = renderer
    .create(
      <Messages
        messages={[
          {
            id: "cd445e6d-e514-424f-ba8f-16ec842002c6",
            userId: "fe27b760-a915-475c-80bb-7cdf14cc6ef3",
            message:
              "Quisque erat eros, viverra eget, congue eget, semper rutrum, nulla.",
            timestamp: "2017-02-09T04:27:38Z"
          }
        ]}
        users={[
          {
            id: "fe27b760-a915-475c-80bb-7cdf14cc6ef3",
            firstName: "Albert",
            lastName: "Rose",
            email: "arosec@bbb.org",
            avatar: null,
            ip: "20.79.231.223"
          }
        ]}
        currentUser=""
      />,
      { createNodeMock }
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

const tree = renderer
    .create(
      <Messages
        messages={[
          {
            id: "cd445e6d-e514-424f-ba8f-16ec842002c6",
            userId: "fe27b760-a915-475c-80bb-7cdf14cc6ef3",
            message:
              "Quisque erat eros, viverra eget, congue eget, semper rutrum, nulla.",
            timestamp: "2017-02-09T04:27:38Z"
          }
        ]}
        users={[
          {
            id: "fe27b760-a915-475c-80bb-7cdf14cc6ef3",
            firstName: "Albert",
            lastName: "Rose",
            email: "arosec@bbb.org",
            avatar: "http://dummyimage.com/100x100.jpg/5fa2dd/ffffff",
            ip: "20.79.231.223"
          }
        ]}
        currentUser="fe27b760-a915-475c-80bb-7cdf14cc6ef3"
      />,
      { createNodeMock }
    ).root
  expect(tree.findByProps({className: "message current-user"}).props.className).toMatch(/current-user/)