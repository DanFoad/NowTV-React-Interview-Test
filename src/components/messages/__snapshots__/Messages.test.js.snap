exports[`test doesn't display message if no user matches 1`] = `
<div
  className="messages" />
`;

exports[`test renders correctly 1`] = `
<div
  className="messages">
  <div
    className="message">
    <div
      className="message-content">
      <div
        className="user-details">
        <img
          alt=""
          className="user-details__avatar"
          src="http://dummyimage.com/100x100.jpg/5fa2dd/ffffff" />
      </div>
      <div
        className="message-details">
        <p
          className="message-details__content"
          title="arosec@bbb.org">
          Quisque erat eros, viverra eget, congue eget, semper rutrum, nulla.
        </p>
      </div>
    </div>
    <div
      className="message-info">
      <p
        className="user-details__name">
        Albert Rose
      </p>
      <p
        className="message-details__timestamp">
        2017-2-9 04:27:38
      </p>
    </div>
  </div>
</div>
`;

exports[`test renders unknownuser avatar for users without avatar 1`] = `
<div
  className="messages">
  <div
    className="message">
    <div
      className="message-content">
      <div
        className="user-details">
        <img
          alt=""
          className="user-details__avatar"
          src="/unknownuser.png" />
      </div>
      <div
        className="message-details">
        <p
          className="message-details__content"
          title="arosec@bbb.org">
          Quisque erat eros, viverra eget, congue eget, semper rutrum, nulla.
        </p>
      </div>
    </div>
    <div
      className="message-info">
      <p
        className="user-details__name">
        Albert Rose
      </p>
      <p
        className="message-details__timestamp">
        2017-2-9 04:27:38
      </p>
    </div>
  </div>
</div>
`;
